### PR TITLE
DSD for TorchTune LoRA

### LINKED ISSUE #128745: DSD for TorchTune LoRA
### 🚀 The feature, motivation and pitch

per discussion with @fegin and Iris, I did a minimal repro of what's needed for TorchTune to use DSD for lora

for full_state_dict=True, the error is 
`# 'aten.copy_.default: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!',).`

for `broadcast_from_rank0=True, full_state_dict=True`, the error is 
`NotImplementedError: c10d::broadcast_: attempted to run this operator with Meta tensors`

```
import logging
import os
from typing import Callable, Optional, Tuple
from torch.distributed._tensor import DTensor

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed.fsdp.wrap import ModuleWrapPolicy
from torch.distributed.checkpoint import state_dict as ptd_state_dict
from torch.distributed._composable.fsdp import fully_shard
import pdb
import sys

def main():
    dist.init_process_group(backend="nccl")
    gpu_id = int(os.environ["LOCAL_RANK"])
    device = f"cuda:{gpu_id}"
    torch.cuda.set_device(device)
    torch.manual_seed(0)
    with torch.device("meta"):
        meta_model = nn.Sequential(*[nn.Linear(4, 4, bias=False) for _ in range(2)])
        for layer in meta_model:
            fully_shard(layer)
        fully_shard(meta_model)
    with torch.device("cpu"):
        cpu_model = nn.Sequential(*[nn.Linear(4, 4, bias=False) for _ in range(2)])
        full_sd = cpu_model.state_dict()
    ptd_state_dict.set_model_state_dict(
        meta_model,
        model_state_dict=full_sd,
        # 'aten.copy_.default: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!',).
        options=ptd_state_dict.StateDictOptions(
            full_state_dict=True, strict=False
        )
        # NotImplementedError: c10d::broadcast_: attempted to run this operator with Meta tensors
        # options=ptd_state_dict.StateDictOptions(
        #     broadcast_from_rank0=True, full_state_dict=True, strict=False
        # )
    )
    
if __name__ == "__main__":
    main()
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mori360**
Fixes #128745
Solve the issue with conflicts when users use full_state_dict while the model is FSDP.

Current solve the issue for `full_state_dict=True`, with error
`'aten.copy_.default: got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!',).`

TODO: for` broadcast_from_rank0=True, full_state_dict=True`, the error is
`NotImplementedError: c10d::broadcast_: attempted to run this operator with Meta tensors`

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @yf225 @chauhang

### CODE REVIEW

**Comment by fegin** [comment]
`broadcast_from_rank0=True` is already implemented. The issue or TODO should be supporting Meta device. Please correct me if I'm wrong.

**Comment by mori360** [comment]
> `broadcast_from_rank0=True` is already implemented. The issue or TODO should be supporting Meta device. Please correct me if I'm wrong.

The PR here aims to solve the issue at https://github.com/pytorch/pytorch/issues/128745 when `broadcast_from_rank0=False`

File: torch/distributed/checkpoint/state_dict.py [hunk 1]
```diff
@@ -561,6 +561,19 @@ def _load_model_state_dict(
         )
         for fqn, local_state in local_state_dict.items():
             state_dict[fqn] = local_state
+    elif info.fsdp_context() and info.full_state_dict:
+        device = None
+        for key, value in local_state_dict.items():
```
Comment (LucasLLC) [inline torch/distributed/checkpoint/state_dict.py:566]: Could we add some comments explaining we expect all Tensors should be on the same device? Especially if the assertion is raised?

Comment (mori360) [reply torch/distributed/checkpoint/state_dict.py:566]: The PR here aim to solve the failure in the copy the tensor value of model_state_dict in the `copy_` function.
The device setting here won't affect the value copy but avoid the failures there.

File: torch/distributed/checkpoint/state_dict.py [hunk 2]
```diff
@@ -561,6 +561,19 @@ def _load_model_state_dict(
         )
         for fqn, local_state in local_state_dict.items():
             state_dict[fqn] = local_state
+    elif info.fsdp_context() and info.full_state_dict:
```
Comment (fegin) [inline torch/distributed/checkpoint/state_dict.py:564]: Why do we care about `fsdp_context()`? This feature should be good for all parallelisms.

**Review by fegin — CHANGES_REQUESTED** [review]

File: torch/distributed/checkpoint/state_dict.py (hunk 2 above)
Comment (mori360) [reply torch/distributed/checkpoint/state_dict.py:564]: I was not sure if the error only appears with 'fsdp', will remove it in the next commit.

File: test/distributed/checkpoint/test_state_dict.py [hunk 3]
```diff
@@ -895,6 +895,48 @@ def init_model_optim():
...
+                fully_shard(layer)
+            fully_shard(meta_model)
+        with torch.device("cpu"):
+            cpu_model = nn.Sequential(*[nn.Linear(4, 4, bias=False) for _ in range(2)])
+            full_sd = cpu_model.state_dict()
+        ptd_state_dict.set_model_state_dict(
+            meta_model,
+            model_state_dict=full_sd,
+            options=ptd_state_dict.StateDictOptions(full_state_dict=True, strict=False),
+        )
```
Comment (fegin) [inline test/distributed/checkpoint/test_state_dict.py:915]: Can we compare the state after `set_model_state_dict()` with the golden sample (cpu_model in this case)?

File: test/distributed/checkpoint/test_state_dict.py [hunk 4]
```diff
@@ -895,6 +895,48 @@ def init_model_optim():
...
+        with torch.device("cpu"):
+            cpu_model = nn.Sequential(*[nn.Linear(4, 4, bias=False) for _ in range(2)])
+            full_sd = cpu_model.state_dict()
+        ptd_state_dict.set_model_state_dict(
+            meta_model,
+            model_state_dict=full_sd,
+            options=ptd_state_dict.StateDictOptions(
+                broadcast_from_rank0=True, full_state_dict=True, strict=False
+            ),
+        )
```
Comment (fegin) [inline test/distributed/checkpoint/test_state_dict.py:938]: ditto.

**Review by fegin — COMMENTED** [review]
LGTM, just need to improve the test cases before landing.

File: test/distributed/checkpoint/test_state_dict.py [hunk 5]
```diff
@@ -895,6 +895,58 @@ def init_model_optim():
...
+            model_state_dict=full_sd,
+            options=StateDictOptions(
+                broadcast_from_rank0=True, full_state_dict=True, strict=False
+            ),
+        )
+        meta_model_dict = meta_model.state_dict()
+        cpu_mocel_dict = get_model_state_dict(cpu_model)
+        for key, value in cpu_mocel_dict.items():
+            meta_model_list = meta_model_dict[key].tolist()
+            self.assertEqual(meta_model_list, value.tolist())
```
Comment (fegin) [inline test/distributed/checkpoint/test_state_dict.py:948]: What do we have to convert it to list? Can't we just compare the values?

File: test/distributed/checkpoint/test_state_dict.py [hunk 6]
```diff
@@ -895,6 +895,58 @@ def init_model_optim():
...
+        set_model_state_dict(
+            meta_model,
+            model_state_dict=full_sd,
+            options=StateDictOptions(full_state_dict=True, strict=False),
+        )
+        meta_model_dict = meta_model.state_dict()
+        cpu_mocel_dict = get_model_state_dict(cpu_model)
+        for key, value in cpu_mocel_dict.items():
+            meta_model_list = meta_model_dict[key].tolist()
+            self.assertEqual(meta_model_list, value.tolist())
```
Comment (fegin) [inline test/distributed/checkpoint/test_state_dict.py:920]: ditto

**Review by fegin — APPROVED** [review]
LGTM, some nits to change before landing.

**Comment by weifengpy** [comment]
discussed offline: follow up is checking memory traces to understand if we are using 2x memory (1x from model.to_emtpy, 1x from SD)

File: test/distributed/checkpoint/test_state_dict.py [hunk 7]
```diff
@@ -895,6 +895,60 @@ def init_model_optim():
...
+        with torch.device("cpu"):
+            cpu_model = nn.Sequential(*[nn.Linear(4, 4, bias=False) for _ in range(2)])
+            full_sd = cpu_model.state_dict()
+        set_model_state_dict(
+            meta_model,
+            model_state_dict=full_sd,
+            options=StateDictOptions(full_state_dict=True, strict=False),
+        )
+        meta_model_dict = meta_model.state_dict()
+        cpu_mocel_dict = get_model_state_dict(cpu_model)
```
Comment (awgu) [inline test/distributed/checkpoint/test_state_dict.py:917]: What is the purpose of calling `get_model_state_dict()` on a module that has no parallelism applied? Is there any difference compared to calling `cpu_model.state_dict()`?

**Pull request closed without merging**